### PR TITLE
pkg/viaduct/lane: add unit tests for lane package

### LINKED ISSUE #6935: pkg/viaduct/lane: add unit tests for WSLane / QuicLane / WSLaneWithoutPack
**What would you like to be added/modified:**

`pkg/viaduct/pkg/lane/` contains the message-framing code used by both WebSocket and QUIC transports (`lane.go`, `ws.go`, `wsnopack.go`, `quic.go`). None of these files has a companion `_test.go`.

Suggested coverage (focus on the WS variants, which can be tested without a real network):

- `NewLane` returns the right concrete type for each protocol and logs/returns `nil` for an unknown protocol.
- `WSLaneWithoutPack.ReadMessage` / `WriteMessage` round trip through a `*websocket.Conn` produced by `httptest`.
- `WSLane` (the `wspack` variant) preserves message payloads across encode/decode.
- `SetReadDeadline` / `SetWriteDeadline` propagate the deadline to the underlying conn.

**Why is this needed:**

Lanes are the actual byte-level encoders/decoders. They sit underneath both `WSConnection` and `QuicConnection` and are referenced (via `lane.NewLane`) in the bug described in Issue 1 above. They're a high-leverage place to add tests because the same code is exercised by every message between cloud and edge.

### PR DESCRIPTION
**Pull request by Priyanshubhartistm**
/kind test

**What this PR does / why we need it**:

pkg/viaduct/pkg/lane/ contains the message-framing code used by both WebSocket and QUIC transports (lane.go, ws.go, wsnopack.go, quic.go). None of these files had a companion _test.go.

This PR adds lane_test.go with 17 unit tests covering:

- NewLane dispatch — returns the correct concrete type for ProtocolTypeWS, ProtocolTypeQuic, and nil for an unknown protocol string
- NewWSLane / NewWSLaneWithoutPack constructors — success path and bad-type path (non-*websocket.Conn argument)
- WSLaneWithoutPack round-trip — ReadMessage / WriteMessage through a real *websocket.Conn produced by httptest.NewServer (no TLS, no external network required)
- WSLane (wspack/protobuf variant) round-trip — ReadMessage / WriteMessage verifying that message ID, parent ID, source, resource, and operation are preserved across encode → decode
- Read/Write deadline expiry — verifies that reads and writes on both WSLane and WSLaneWithoutPack return errors once a deadline has passed (using real network connections with bounded waits)
- SetReadDeadline / SetWriteDeadline — covers the zero-time case (clearing the deadline)
- Error paths — ReadAfterClose, WriteAfterClose, and ReadMalformedData for WSLaneWithoutPack; malformed protobuf read for WSLane

**Which issue(s) this PR fixes**:

Fixes #6935

## Test Evidence

**Screenshot:**
image

$ go test -race ./pkg/viaduct/pkg/lane/...
ok      github.com/kubeedge/kubeedge/pkg/viaduct/pkg/lane    1.268s

### CODE REVIEW

**Comment by Priyanshubhartistm** [comment]
Hi @DoisLONG,  could you please take a look at this PR?
whenever you have sometime, I would really appreciate your review and any feedback or suggestions you may have.

Thank you for your time!

**Comment by DoisLONG** [comment]
Thanks for adding coverage to this package. The constructor and successful WebSocket round-trip tests are a useful starting point, and the previous goroutine assertion issue has been corrected.

There are still several important gaps in relation to #6935.

### 1. Avoid leaking the upgraded HTTP handler goroutine

`newTestWSPair` keeps the handler blocked on:

```go
<-r.Context().Done()
```

After the WebSocket connection is hijacked, the request context is not an appropriate lifecycle signal for the WebSocket connection. The handler can remain blocked after both WebSocket connections are closed, leaking one goroutine for every helper invocation.

The handler can return immediately after handing the upgraded connection to the test; the hijacked WebSocket remains valid after `ServeHTTP` returns.

Please also use a connection/error channel rather than waiting only on `ready`, so an upgrade failure cannot leave the test waiting indefinitely.

### 2. Cover the QUIC branch of `NewLane`

Issue #6935 explicitly asks for `NewLane` to return the correct concrete type for each supported protocol. The current tests cover only `ProtocolTypeWS` and an unknown protocol, leaving the `ProtocolTypeQuic` branch untested.

A real QUIC network is not required for this constructor test. A minimal fake implementing `quic.Stream` can be passed to `NewLane`, and the result can be asserted as `*QuicLane`.

Please also update the PR notes: the issue focuses on WS round trips, but it does not exclude the `NewLane` QUIC dispatch test.

### 3. Verify the actual message payload

Both round-trip tests construct messages without `Content` and assert only routing/header fields. They would still pass if content encoding and decoding were removed.

Please add non-empty content and verify it after the round trip:

* a JSON-safe string or object for `WSLaneWithoutPack`;
* a string or `[]byte` payload for the packed `WSLane`.

Please also verify `Group`, `Timestamp`, and `Sync`, since these are part of the message conversion contract and `Sync` is copied separately by the translator.

### 4. Verify deadline propagation behavior, not only private fields

The deadline tests currently verify only that `l.readDeadline` or `l.writeDeadline` was assigned. They do not prove that the deadline was applied to the underlying `websocket.Conn`.

These tests would continue passing if the calls to `conn.SetReadDeadline` and `conn.SetWriteDeadline` were removed.

Please test an observable behavior, for example by setting an expired read deadline and verifying that a read returns a timeout promptly. Apply the same principle to write-deadline coverage, or introduce an appropriate injectable connection abstraction.

### 5. Add focused error-path coverage

Please cover at least:

* a read or write after the peer/connection is closed;
* malformed packed data causing `WSLane.ReadMessage` to return an error.

These are important branches in the lane's transport and decoder behavior.

After the updates, please provide copyable text output for:

```bash
go test -race ./pkg/viaduct/pkg/lane/...
```

I would like to review the updated helper and the final behavioral coverage.

**Comment by Priyanshubhartistm** [comment]
Hi @DoisLONG , could you please take a look at this PR when you have time? I’ve addressed the review feedback and would really appreciate your feedback.

**Comment by DoisLONG** [comment]
Thanks for the update. The main review items have been addressed: the WebSocket helper no longer leaks its HTTP handler, the QUIC constructor branch is covered, non-empty payloads are verified, and focused error paths have been added.

I found two remaining test-correctness details.

### 1. Verify the message timestamp

Both round-trip messages set `Header.Timestamp`, but neither test asserts that the decoded message preserves it.

Please add a timestamp assertion to both the packed and non-packed round-trip tests. Otherwise, dropping the timestamp during translation would not fail the tests.

### 2. Bound the expired-read-deadline tests

The expired read-deadline tests call `ReadMessage` synchronously. If deadline propagation regresses, the test will block until the package-level `go test` timeout rather than fail promptly.

Please run the read in a goroutine and use a bounded `select`, for example around one second. On timeout, close the connection and fail the test.

Since these tests specifically verify deadline expiry, please also require the returned error to satisfy the timeout interface and report `Timeout() == true`, rather than accepting any non-nil error.

File: pkg/viaduct/pkg/lane/lane_test.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+// the server-side conn and the client-side conn.  Both are closed when t finishes.
+func newTestWSPair(t *testing.T) (serverConn, clientConn *websocket.Conn) {
+	t.Helper()
+
+	var sConn *websocket.Conn
+	ready := make(chan struct{})
+
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c, err := wsUpgrader.Upgrade(w, r, nil)
+		require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (Priyanshubhartistm) [reply pkg/viaduct/pkg/lane/lane_test.go]: done

**Comment by Priyanshubhartistm** [comment]
Hey @DoisLONG , Done! lmk if any other changes are needed 
thanks

**Comment by DoisLONG** [comment]
Thanks for the update. This looks much better now.

The round-trip tests now verify the important message fields, including timestamp and sync flag, and the deadline tests exercise real expired-deadline behavior with bounded waits. The WebSocket test helper also avoids calling `require.*` from the HTTP handler goroutine, which addresses the earlier testing-contract concern.

I do not see any blocking issue from my side.

A few non-blocking suggestions:

* It would be nice to add the same expired write-deadline coverage for `WSLane`, since the current write-deadline test only covers `WSLaneWithoutPack`.
* `TestWSLaneWithoutPack_ReadAfterClose` currently uses a short `time.Sleep`; setting a short read deadline before `ReadMessage` would make the test more deterministic.
* Please make sure the CI workflows are triggered and pass before merge.

**Comment by Priyanshubhartistm** [comment]
@DoisLONG, thank you for the review! Addressed your suggestions

Added TestWSLane_WriteDeadlineExpired for the packed WSLane variant.
Replaced time.Sleep in TestWSLaneWithoutPack_ReadAfterClose with SetReadDeadline(200ms) for determinism.
All 17 tests pass

**Comment by DoisLONG** [comment]
Thanks for adding these lane tests.

I checked the current version, and the earlier inline comments around the WebSocket test helper seem to have been addressed. The real WebSocket round-trip tests for both `WSLaneWithoutPack` and `WSLane`, together with the deadline/error-path cases, are useful coverage for this package.

One small non-blocking suggestion: the PR description says it adds 13 unit tests, but the current test file seems to include more cases now. Could you update the description to match the current scope?

Other than that, this looks reasonable to me.

**Comment by Priyanshubhartistm** [comment]
Hi @DoisLONG, updated the PR description to reflect the current count of 17 tests. Thanks for the  review!

File: pkg/viaduct/pkg/lane/lane_test.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	url := "ws" + strings.TrimPrefix(srv.URL, "http")
+	cConn, _, err := websocket.DefaultDialer.Dial(url, nil)
+	require.NoError(t, err)
+
+	<-ready // wait until server has upgraded
+
+	t.Cleanup(func() {
+		_ = cConn.Close()
+		_ = sConn.Close()
+	})
```
[the remark this answers is not in the document]
Comment (Priyanshubhartistm) [reply pkg/viaduct/pkg/lane/lane_test.go:99]: good catch

**Comment by Priyanshubhartistm** [comment]
Thank you for the review, @DoisLONG , @WillardHu
If everything looks good now, could you please /approve this PR when you have a chance? I'd really appreciate it.
